src/ray/object_manager/pull_manager.cc

```diff
@@ -96,30 +98,37 @@ void PullManager::TryToMakeObjectLocal(const ObjectID &object_id) {
     return;
   }
 
+  // New object locations were found, so begin trying to pull from a
+  // client. This will be called every time a new client location
+  // appears.
+  bool did_pull = PullFromRandomLocation(object_id);
+  if (did_pull) {
+    UpdateRetryTimer(request);
+    return;
+  }
+
+  // If we cannot pull, try restoring objects from the external storage.
   if (!request.spilled_url.empty()) {
-    // Try to restore the spilled object.
+    const auto spilled_node_id = request.spilled_node_id;
+    RAY_CHECK(!spilled_node_id.IsNil());
```

Comment:
Isn't this check too strong? I thought this would be nil for other storage systems, like S3.

Reply:
It depends on if we'd like to keep the previous protocol for different types of external storages. I'd prefer to keep the consistent implementation for all, but if you think it's better keeping the previous protocol for distributed storages, I can do that.

Reply:
Oh, also the node id is added in `AddSpilledUrls`, so it should always be guaranteed to have node id (since we set the node id there).